zenith_utils/src/postgres_backend.rs

```diff
@@ -0,0 +1,147 @@
+// Server-side Postgres connection, as limited as we need.
+
+use crate::pq_proto::{BeMessage, FeMessage, FeStartupMessage, StartupRequestCode};
+use anyhow::bail;
+use anyhow::Result;
+use bytes::Bytes;
+use log::*;
+use std::io;
+use std::io::{BufReader, BufWriter, Write};
+use std::net::TcpStream;
+
+pub trait Handler {
+    // handle single query
+    fn process_query(&mut self, pgb: &mut PostgresBackend, query_string: Bytes) -> Result<()>;
+    // called on startup packet receival, allows to process params
+    fn startup(&mut self, _pgb: &mut PostgresBackend, _sm: &FeStartupMessage) -> Result<()> {
+        Ok(())
+    }
+}
+
+pub struct PostgresBackend {
+    socket: TcpStream,
+    // replication.rs wants to handle reading on its own in separate thread, so
+    // wrap in Option to be able to take and transfer the BufReader. Ugly, but I
+    // have no better ideas.
```

Comment:
Would it make sense to split this into a reader/handler and a writer?

Reply:
Hm, what do you mean?

Reply:
We could have two separate structs for reading and writing postgres streams. Then `replication.rs` can consume the reader type and access the `BufReader` directly

Reply:
Well, I think carrying two structs everywhere is not worth it, and once we pack the inside one we get the same problem.